src/dataflow/utilities/aws.ts

```diff
@@ -114,9 +129,49 @@ export const fetchProgramData = (programId: string, time?: number, endTime?: num
       if (error) {
         reject(error);
       }
-      if (data) {
-        resolve(JSON.parse(data.Payload as string));
+      if (data && data.Payload) {
+        const result: any = { count: 0 };
+        let payload: any;
+        try {
+          payload = JSON.parse(data.Payload as string);
+        } catch (e) {
+          // In case of strange messages in MQTT, handle a glitch gracefully
+          resolve(result);
+        }
+        if (payload.data) {
+          // Add latest data to cache
+          addDataPoint(programId, JSON.parse(data.Payload as string));
+          result.count = allProgramsCache[programId].data ? allProgramsCache[programId].data.length : 0;
+          result.data = allProgramsCache[programId].data;
+        }
+        resolve(result);
       }
     });
   });
 };
+
+const addDataPoint = (programId: string, payload: any) => {
```

Comment:
```suggestion
const addDataPoints = (programId: string, payload: any) => {
```
since it adds multiple points